Implementation of StreamSizeScheduler and Tests

jguery opened the pull request:
This does not yet contain a persistent store for the stream size schedules.

Jira: https://issues.cask.co/browse/CDAP-853
Build: http://builds.cask.co/browse/CDAP-DUT833-1

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java:
```diff
@@ -73,7 +73,11 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
-      taskRunner.run(Id.Program.from(accountId, applicationId, programId), programType, args);
+      try {
+        taskRunner.run(Id.Program.from(accountId, applicationId, programId), programType, args);
+      } catch (TaskExecutionException e) {
```
is there a reason you need a different exception?

jguery replied:
Yes, there is. I want to use the ``ScheduleTaskRunner`` to trigger job execution in any schedule. The ``run`` method used to throw a quartz exception if the execution failed. It now throws a ``TaskExecutionException``, which all schedules can understand. The class in which this code is executed though, ``ScheduledJob``, implements a quartz interface which throws a ``JobExecutionException``.

anew replied:
I see. I did not realize that it was a Quartz exception.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -41,42 +68,97 @@
+    schedule(program, programType, streamSizeSchedule, true, -1, -1, true);
+  }
+
+  private void schedule(Id.Program program, SchedulableProgramType programType, StreamSizeSchedule streamSizeSchedule,
```
this deserves a javadoc - what is baseRunSize, what is baseRunTs?

jguery replied:
I will add. baseRunSize is the size from which a schedule starts counting an increment - which is likely to be the size that we had the last time the schedule was triggered. baseRunTs is the time at which that baseRunSize was obtained.

anew replied:
And what is Ts? Why not call it Time?

jguery replied:
Ts is for timestamp.

anew replied:
Is this ever called with parameters other than true, -1, -1, true? I can't see it.

jguery replied:
No right now it isn't, but once we have a persistent store for the schedule tasks, it will be called with whatever is stored in the store.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -41,42 +68,97 @@
+        }
+
+        // Add the scheduleTask to the StreamSubscriber
+        streamSubscriber.createScheduleTask(program, programType, streamSizeSchedule,
```
what happens if there is an existing subscriber for the same schedule?

jguery replied:
Then the ``streamSubscriber.createScheduleTask`` method won't replace the existing schedule. I should make this method return ``true`` when we do add a new schedule, so that the next line, 
```
scheduleSubscribers.put(getScheduleId(program, programType, streamSizeSchedule.getName()), streamSubscriber);
```
is executed only if the call returned ``true``.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -86,32 +168,81 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    String programScheduleId = getProgramScheduleId(program, programType);
+    for (String scheduleId : scheduleIds) {
+      if (scheduleId.startsWith(programScheduleId)) {
+        builder.add(scheduleId);
```
why copy twice?

anew replied:
If it is an ordered concurrent map, then you could also do copyOf(submap.keySet())

jguery replied:
Ok will change that, and the map will be a ``ConcurrentSkipListMap``.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -86,32 +168,81 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
   public List<String> getScheduleIds(Id.Program program, SchedulableProgramType programType) {
-    return ImmutableList.of();
+    Set<String> scheduleIds;
+    synchronized (this) {
```
not sure why this needs to be synchronized. Isn't it a concurrent map?

jguery replied:
It will get removed with the changes from above.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -86,32 +168,81 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    StreamSubscriber subscriber = scheduleSubscribers.remove(getScheduleId(programId, programType, scheduleName));
+    if (subscriber != null) {
+      subscriber.deleteSchedule(programId, programType, scheduleName);
+      synchronized (this) {
```
I am not sure I understand. A subscriber can multiple tasks? You only cancel the subscriber, if this was the only task for that subscriber. But if it has other tasks, then it is already removes from the scheduleSubscribers... So there is no reference to it anymore, but it has more active tasks?

jguery replied:
I'm not sure I understand your question either. Yes a subscriber can have many tasks. One subscriber is responsible for listening to the notifications of one stream, and to poll the size of that stream. Many schedules can be attached to the same stream, hence one subscriber can have many schedule tasks. The subscriber will just pass the notifications along to the schedule tasks, which keep a state and know exactly when they should trigger the execution of a job.
There are two maps to store those objects in the StreamSizeScheduler class: ``scheduleSubscribers``, and ``streamSubscribers``. 
``streamSubscribers`` stores for each stream, its dedicated subscriber.
``scheduleSubscribers`` stores for each schedule ID, its subscriber, which is the same object as referenced by the previous ``streamSubscribers`` map. This map is needed because most APIs of the ``Scheduler`` interface only allow us to construct a scheduleId, which does not contain the stream name.
So when deleting a schedule, it is safe to always delete the entry from the ``scheduleSubscribers`` map, but we can only delete the entry from the ``streamSubscribers`` map if there is no more schedule attached to the stream. This is why there is synchronization in the ``deleteSchedule`` and ``schedule`` methods, to protect the ``streamSubscribers`` map. But this might be done in a wrong way, let me know if you see something wrong.

anew replied:
ok, got it. Personal discussion helped a lot :)

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -86,32 +168,81 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
   public void deleteSchedules(Id.Program programId, SchedulableProgramType programType) {
-
+    Set<String> scheduleIds;
+    synchronized (this) {
```
again it is not clear how synchronizing helps here. If you don't synchronize the following code, then the subscribers can still get updated concurrently.

jguery replied:
Ok, I'll use the ConcurrentSkipListMap here, like mentioned above.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
```
Can this be a separate file? Or does it have to be an inner class?

jguery replied:
It doesn't have to be, but since this is very internal to the ``StreamSizeScheduler``... In any case, if I have to move it, I'll move it only after you reviewed my changes completely, since it would make it hard to see what's changed.

anew replied:
I just like small files. But you can keep it in here.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
```
What happens if this fails? This would throw an exception that pops up to some caller trying to create a schedule? Is there a retry logic? And if this succeeds, can it happen that the subscription is interrupted by some failure, will it reconnect?

jguery replied:
If it succeeds, I am not sure if the subscription can be interrupted. The notification service relies on Twill's ``KafkaClient`` to subscribe to the notifications in Kafka. I don't know how Twill handles possible interruption in the subscription, or if it is even possible. @chtyim, what is your take on this?

It could fail for 2 reasons: 
- if the stream does not exist, then a NotificationFeedNotFoundException will be thrown. I think it's a legitimate error, which should revert the creation of the schedule. When deploying an app that has one such schedule, the exception should be caught and logged, but the app should still be deployed.
- if the notification service cannot run, for eg because the dataset service is not up, then a NotificationFeedException will be thrown. In that case we should retry, you're right. But what should define how long we retry?

jguery replied:
Also, there could be some inconsistency in the app deployment phase. So far I think that we considered that creating a schedule could never fail, and so there is no exception handling when creating the schedules, when deploying an app. But now, this new schedule can fail if the stream it references doesn't exist. 

Yet, with the current behavior, 1. if one schedule fails to be created, then we don't even try to create the remaining schedules, and 2. the app is still deployed, and the ApplicationSpecification references a schedule which could not be started. What should be done here? 

If we don't want to make the app deployment fail when a stream doesn't exist, then I think we will have to wait until that JIRA - https://issues.cask.co/browse/CDAP-1446 - gets resolved. Basically, it means that the schedule would be created, but not started. The schedule should wait till the stream is created to start.

anew replied:
As discussed in person, let's require for now that he stream must exist (or it is created by the app that contains the schedule) when the schedule is created.

jguery replied:
Yup.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+                                  newTask);
+      if (previous != null) {
+        // We cannot replace an existing schedule - that functionality is not wanted - yet
+        return;
```
so we silently do nothing? Don't you have to return an error? Otherwise the caller thinks it was successful

anew replied:
what happens in the case where an app gets redeployed with a new schedule? Are the old schedules canceled first? And then recreated?

jguery replied:
Here I will return a boolean to say whether the creation was successful, or if the schedule already existed, and I will log a warning.
When redeploying an app, yes the schedules are deleted first and then recreated.

anew replied:
I think this discussion is obsolete with your recent refactoring of schedule management into the app deploy pipeline.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+
+      // Initialize the schedule task
+      if (baseRunSize == -1 && baseRunTs == -1) {
+        // This is likely to be the first time that we schedule this task - ie it was not in the schedule store
```
likely? And in the unlikely event that it is not he first time?

jguery replied:
It should actually never happen because when we read the store, it will never contain the values -1, instead the values for baseTs and baseSize will be the ones set in the lines just below (but that is only coming in a next PR).
But if it does happen, for some reason, then the same logic still applies.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      long pastRunTs;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen when a stream is truncated: the baseSize is still the old size,
```
even if the stream was truncated, there can still be enough new data to trigger the schedule. I don't think it is correct to return here.

jguery replied:
Yes, that's correct. Although, the truncation logic will disappear once https://github.com/caskdata/cdap/pull/1364 is merged, and the chances to take it into account are done. That is because the notifications will copy the metrics, and give the size of all the events ever ingested by the stream, and no truncation will be handled.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        if (notification.getSize() < baseSize) {
+          // This can happen when a stream is truncated: the baseSize is still the old size,
+          // but we receive notification with way less data
+          baseSize = notification.getSize();
```
Also, I thought notifications are based on total bytes written to stream. not current stream size.

jguery replied:
Nope it wasn't the case. It will be soon.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java:
```diff
@@ -38,4 +38,10 @@
   public static final String SCHEDULE_NAME = "scheduleName";
```
can this be schedule.name? And with namespaces (do we have that yet?), would this contain the namespace? Or is it always the schedule name, because that is within the scope of the application?

jguery replied:
I think this constant was introduced in 2.7, and users have access to it. Is it okay to change it? It does not contain the namespace, which will always be the namespace of the application, yes.

anew replied:
I guess I am just a little irritated that even for constants added in the same PR, we cannot be consistent (see the next couple of lines).

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.running = active;
```
can you name this active instead of running? "running" suggests that something is actually running.

jguery replied:
Okay.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        ProgramOptionConstants.PAST_RUN_DATA_SIZE, Long.toString(pastRunSize)
+      ));
+
+      while (true) {
```
you retry infinitely?

anew replied:
I think it would be better to retry only a few times. And if that fails, reset base time and size to what it was before (since they reflect the last run of the schedule).

anew replied:
And this whole indefinite retry loop is inside a synchronize ?

anew replied:
The interesting thing here is what happens if another run gets scheduled while the previous one is in its retry?

jguery replied:
The synchronize block is only to update the baseSize and baseTs, it doesn't contain the retry loop. If another run gets scheduled, then it could run concurrently with the first one, and the program options will reflect the state of the run.
Do we really want to make it retry only a few times? For now the re-execution depends on the `refireImmediatly` flag that is in the ``TaskExecutionException``, and that reproduces the behavior in the Quartz scheduler.

anew replied:
So you would retry indefinitely? I think that is wrong. But I am not sure what sets the refireImmediately flag. I think the number of retries should be bounded, and if the time scheduler does not do that, then it should be fixed, too.

jguery replied:
I checked in the ``TimeScheduler`` and we are actually responsible for throwing the ``TaskExecutionException`` and setting the refireImmediately flag, and we always set it to false. This happens in the ``ScheduleTaskRunner`` class.
So we actually only try execution once.

anew replied:
Let's remove the loop then.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+    /**
+     * @return size of the stream queried directly from the file system
+     */
+    private long pollStream() {
```
this should be done using metrics, not file system scanning.

jguery replied:
This will come after this PR https://github.com/caskdata/cdap/pull/1364 is merged.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
```
persist is ignored?

jguery replied:
For now it is, until the persistent store is implemented (coming in a future PR).

anew replied:
ok. looking forward to that PR

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+          lastNotification = new StreamSizeNotification(baseTs, baseSize);
+        }
+      } else {
+        newTask.startSchedule(baseRunSize, baseRunTs, active, persist);
```
It appears that you assume this: if (baseSize or baseRunTs not -1) then lastNotification is set to something non-null. Is that always true? I am not sure understand the correlation between when baseRunSize and baseRunTs are passed in and != null, and whether the last notification is set.

anew replied:
Or can it happen that a task is scheduled, lastBaseSize is != -1, but lastNotification is null? In that case, we would not call received() in lime 337. But the comment there states that we always want to initialize a new task with the current state of the stream size.

jguery replied:
You're right in your second comment. Typically, when we have the persistent store, this will happen: a schedule will be loaded from the persistent store, with values of baseRunSize and baseRunTs != -1 - they will be the values of the last time a job was triggered, or the values of the initialization of the schedule, if no job was ever triggered. Yet, lastNotification will be null, and there is nothing to do here but to wait for the next notification published by the stream.

The comment misled you then. Which comment are you talking about? "Initialize the schedule task" ? We do initialize the task but it doesn't have to be with the current state of the stream size, it can be with the last recorded state.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        // we want to pass it the last seen notification
+        // This call will send the notification to all the active tasks. The ones which are active before
+        // that method is called will therefore see the notification twice. It is fine though.
+        received(lastNotification, null);
```
Isn't it sufficient to notify only this task? Is there an advantage of notifying all tasks, every time a new task is scheduled?

jguery replied:
There is a case when we update the ``lastNotification``, in which case it's also worth it to notify the other tasks.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+    /**
+     * @return true if we successfully suspended the schedule, false if it was already suspended
+     */
+    public boolean suspend() {
```
should this be synchronized or use an atomic? If two threads call this concurrently, they may both return true. I believe that it may get called concurrently.

jguery replied:
You're right, I should use an AtomicBoolean here.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        return;
+      }
+      synchronized (this) {
+        if (task.suspend()) {
```
better use an AtomicBoolean than synchronize?

jguery replied:
Right, if ``task.suspend()`` is thread safe, then we don't need the synchronized block here.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+     */
+    public void resumeScheduleTask(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null || !task.resume()) {
```
similar to other places. Simply returning here makes the caller believe that resume was successful. But isn't it an error to resume something that does not exist?

anew replied:
Also, task.resume() must be synchronized or use an Atomic. Otherwise two threads may attempt resume concurrently and both believe they succeeded. Then you would increment the active tasks twice etc.

jguery replied:
It is not clear from the ``Scheduler`` interface what is the expected behavior when performing operations on an inexistent schedule. The time scheduler, which internally uses a quartz scheduler, has an inconsistent behavior. Resuming or suspending an inexistent schedule does not do anything, whereas deleting one throws a NPE. 
Those situations in practice don't happen, because we check at a higher level in app fabric that a schedule exists. But still, I think that you're right, and at least a ScheduleNotFoundException should be created and thrown in those cases.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -41,42 +68,97 @@
   public StreamSizeScheduler(CConfiguration cConf, NotificationService notificationService, StreamAdmin streamAdmin,
                              StoreFactory storeFactory, ProgramRuntimeService programRuntimeService,
                              PreferencesStore preferencesStore) {
+    this.pollingDelay = cConf.getLong(Constants.Notification.Stream.STREAM_SIZE_SCHEDULE_POLLING_DELAY);
```
better convert this to milliseconds once, than converting it every time you need it.

anew replied:
Is it sufficient to configure the same polling delay for all streams? Or should it be possible to configure that per stream? Or per schedule?

jguery replied:
I will convert to millis only once, that's true.
We talked about that with @chtyim, and decided to only go with a global delay for now, and maybe later enable to configure a delay per schedule.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+            (lastNotification.getTimestamp()
+              + TimeUnit.SECONDS.toMillis(pollingDelay) <= System.currentTimeMillis())) {
+            long streamSize = pollStream();
+            lastNotification = new StreamSizeNotification(System.currentTimeMillis(), streamSize);
```
I am not sure how much it matters, but this is not truly accurate. The streamsSize represents a time that is slightly earlier than now. Actually, if you poll using metrics, it can be delayed by seconds or sometimes tens of seconds. The correct time stamp to use would be the timestamp of the metric. So here that time can diverge a little. Most of the times it won't matter. But even if you keep this logic (to use currentTimeMillis) it would be cleaner if pollStream() returned the time associated with the size that it reports.

jguery replied:
Okay, I will make pollStream() return a timestamp too.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+     */
+    public void deleteSchedule(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask scheduleTask = scheduleTasks.remove(getScheduleId(programId, programType, scheduleName));
+      if (scheduleTask != null && scheduleTask.isRunning()) {
```
you may have to be careful. Suppose a task is suspended, but concurrently being resumed. Then the resuming thread will not know that this thread is deleting. It will hence increment the activeTasks by one, but the task is deleted. The consequence is (I believe) that we will never stop polling.

jguery replied:
That's correct... Another situation is: we are deleting an active schedule task, and creating the same task - with the same id - concurrently. The thread removing can remove the task from the ``scheduleTasks`` map, then the creation thread puts the task back in the map and increments the activeTasks counter. The deleting thread now decrements the activeTasks counter. The counter is off by minus one and polling will never be active if there are no other task for the subscriber.

So it seems like all the methods which modify the scheduleTasks map and the activeTasks counter should do so in a synchronized way. That means deleteSchedule(), suspendScheduleTask(), resumeScheduleTask() and createScheduleTask() should be synchronized, or parts of them should be.

anew replied:
right.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        }
+      }
+      if (send) {
+        sendNotificationToActiveTasks(notification);
```
if there are multiple threads competing here, then only one of them needs to cancel and reschedule the polling, right?

jguery replied:
Yes, you're right, I will move the ``cancelPollingAndScheduleNext`` method into this condition block.

jguery replied:
Actually, putting this method in the condition block may cause it to not be called, and the polling could be not rescheduled.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+    private void sendNotificationToActiveTasks(final StreamSizeNotification notification) {
+      List<StreamSizeScheduleTask> values;
+      synchronized (this) {
+        values = ImmutableList.copyOf(scheduleTasks.values());
```
If the scheduleTasks are a concurrent map, then it is safe to iterate directly over the map. I think it is also better: Because if there is really a concurrent modification (say, a task is deleted) then you want to see that modification reflected here (otherwise you would call a deleted task).

jguery replied:
Ok, I will do that then.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+
+    @Override
+    public void received(StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only pass the stream size notification to the schedule tasks if the notification
```
It seems that notificationContext is never used. What is it for?

jguery replied:
We can use it to access datasets within a transaction. But we don't need that functionality here.

jguery commented:
@anew I addressed your comments, with replies or code changes. Please have another look. Thank you

jguery commented:
Since https://github.com/caskdata/cdap/pull/1364 has been merged, I updated this PR to use metrics instead of the file system to retrieve the size of a stream.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java:
```diff
@@ -92,7 +93,7 @@ protected final void startScheduler() {
-      LOG.info("Stopped stram size scheduler");
+      LOG.info("Stopped stream size scheduler");
     } catch (Throwable t) {
       LOG.error("Error stopping stream size scheduler {}", t.getCause(), t);
```
why does this log t.getCause()? Can't it just log t?

jguery replied:
I kept the old code, but will change.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java:
```diff
@@ -169,6 +171,7 @@ public void resumeSchedule(Id.Program program, SchedulableProgramType programTyp
       scheduler.resumeSchedule(program, programType, scheduleName);
     } catch (NotFoundException e) {
       LOG.trace("Could not resume schedule", e);
+      Throwables.propagate(e);
```
Why propagate this as a runtime exception? What is wrong with re-throwing the NotFoundException?

jguery replied:
The methods of the ``Scheduler`` interface don't throw any exceptions, that's why. I put a comment above, asking if you thought I should modify the interface to better handle errors.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java:
```diff
@@ -179,6 +182,7 @@ public void deleteSchedule(Id.Program program, SchedulableProgramType programTyp
       scheduler.deleteSchedule(program, programType, scheduleName);
     } catch (NotFoundException e) {
       LOG.trace("Could not delete schedule", e);
+      Throwables.propagate(e);
```
This method catches, logs and propagates exceptions, but the following method does not. Why?

jguery replied:
Because the following method doesn't throw a ``NotFoundException``. In the following method (deleting all schedules for one program), we have no way of knowing if the program doesn't exist, or if there is no schedule for an existing program, hence we can't throw that exception.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java:
```diff
@@ -209,12 +213,12 @@ private Scheduler getSchedulerForSchedule(Id.Program program, SchedulableProgram
     Map<String, ScheduleSpecification> schedules = appSpec.getSchedules();
     if (schedules == null || !schedules.containsKey(scheduleName)) {
-      throw new NotFoundException("schedule", scheduleName);
+      throw new ScheduleNotFoundException(scheduleName);
```
In this whole class (and in general, in this new schedule code), exceptions are not handled consistently. Sometimes we throw something meaningful (like a ScheduleNotFoundExn), sometimes, we propagate any exception as a RuntimeException. Sometimes we log, sometimes we don't. You would not think that this code was written by one person.

jguery replied:
Like I said above, the ``Scheduler`` interface doesn't throw any exception. Should that be changed?

jguery commented:
I merged https://github.com/caskdata/cdap/pull/1426 in this PR - the since should reduced once this gets merged. @anew, I addressed your comments. Could you have another look? Thank you.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,43 +196,116 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    return store;
   }
 
   private String getScheduleId(Id.Program program, SchedulableProgramType programType, String scheduleName) {
```
I would call this scheduleIdFor() and make it static.

jguery replied:
Ok will do.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,43 +196,116 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
   @Override
   public List<String> getScheduleIds(Id.Program program, SchedulableProgramType programType)
```
Do you know what this method is good for? Nobody ever seems to call it. I am asking because I am wondering whether that caller would expect repeatable read?

jguery replied:
I'm wondering too, we only use it in tests...

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,43 +196,116 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      } catch (NotFoundException e) {
+        // It can happen that the schedule was deleted while being updated. In which case, the update action
+        // came first and we still want to create it
+        // TODO is it really the behavior that we want here?
```
I don't think this can happen, because all this happens under a single transaction. the fact that the subscriber for this schedule was there means that it must be there, and under snapshot isloation, you won't see a change in the same transaction. I almost feel that this would be an internal error. For now I think it is ok to keep this code as is, but change the log level to warning, so that we see it in the logs if it happens.

jguery replied:
Alright will change the log level.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,43 +196,116 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        deleteSchedule(programId, programType, scheduleId.substring(scheduleIdIdx));
+      } catch (ScheduleNotFoundException e) {
+        // Could be a race, the schedule has just been deleted
+        LOG.debug("Could not delete schedule '{}'", scheduleId, e);
```
same here, should be a warning.

anew replied:
and it should say what the comment says

jguery replied:
Ok.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen if no notification is received for a stream for some time, and we poll the stream events
+          // size using metrics, and some metric events have hit their TTL. In that case it's impossible to know
```
this is scary. Are you saying that because of metrics TTL, stream metrics will shrink over time? I would hope that the total bytes written to a stream is a metric that can only go up.

The only way that this should be possible is if the stream was deleted and recreated. Then, the metrics would start over at 0. @abaranau can you confirm that?

anew replied:
In that case (recreation of stream), the next time the schedule runs, shouldn't it see all data in the new stream from the beginning? That is, is the new size is greater than the threshold, we should run the schedule action; if not we would set baseSize = 0, and baseTs remains what it was before. But I am not sure. This is an intricate case.

anew replied:
Or, if the stream admin emits a notification with size=0 after the stream is recreated, then your current logic works.

anew replied:
Is this also possible if there is a race between polling and a notifcation? Say you poll, get a size x. Now notification comes in with size x+10. It triggers the action. Then the poll calls received() and now you get a smaller size. I guess in that case, we would want ignore the polling, right?

jguery replied:
For your last comment, if there was a race, then the notification containing the polling would have a timestamp lower than the last notification. But the received method here would never be called with such a notification - we make sure of that in the StreamSubscriber#received. Yet, now that I think of it, the polling could indeed show a timestamp higher - ie more recent - but a lower size, because the metrics shows a state which could be a few moments earlier than the timestamp we attach it with. This is because the way we query metrics, we can't retrieve the timestamp of the last seen event - I checked with @gsps1, but will make sure again. In that situation, I don't know what should be done. The better option would be to open a Jira for the metrics issue, like you suggested below. @gsps1 told me that it was supposed to be that way, though.

jguery replied:
For the TTL issue, this is what I understood:
As long as at least one event is ingested by the stream in the period of the TTL, then the counter will never decrease. The counter could suddenly drop to 0 if no event has been ingested in TTL time. I'll double check with @abaranau.
This could lead to some tricky situations: if the stream already has 1.5 GB of data, and no data is ingested in TLL time, the metrics could suddenly show 0, yet the notification counting logic is unaware of that. If 0.5GB are ingested, the notification logic will send: 2GB, but the metrics will show 0.5GB. The polling and the notifications will never be in sync again, until CDAP is restarted and the notifications start counting from the size given by the metrics again.

abaranau replied:
While there's no doubt we want to figure out metrics data expiration (there's https://issues.cask.co/browse/CDAP-1124 started while ago that is related), it can be pretty tricky for dependent components to handle that properly. The biggest complexity with "totals", like in this case. On metrics system side, these will never be TTLed for now: https://issues.cask.co/browse/CDAP-1626.

At the same time, you may also think about handling such cases as metrics data cleanup (manual). I can see, if metrics data is pollutted and takes lots in the cluster, some may decided to manually drop it.. It is another question if we support that explicitly, but the system should not fail completely:)

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+          // we can do.
+          baseSize = notification.getSize();
+          baseTs = notification.getTimestamp();
+          return;
```
don't you want to log this? Similarly to below, the baseSize/Ts was updated. And since this is an uncommon case, I fell that it deserves a warning.

jguery replied:
Right, I will log this.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        ProgramOptionConstants.SCHEDULE_NAME, currentSchedule.getName(),
+        ProgramOptionConstants.LOGICAL_START_TIME, Long.toString(baseTs),
+        ProgramOptionConstants.RUN_DATA_SIZE, Long.toString(baseSize),
+        ProgramOptionConstants.PAST_RUN_LOGICAL_START_TIME, Long.toString(pastRunTs),
```
Is this always the logical start time of the last run? Or or is this the time that corresponds to the base size that was used to compute the amount of new data? Due to the special case above (and also if a run fails), it could be that there was no run at that time.

jguery replied:
You're right, it is actually the time corresponding to the base size. It's actually also the case for ``LOGICAL_START_TIME``, it's the time at which we received the new data that triggers the new execution, and not the time at which we start the execution... Should I use different constants?

jguery replied:
Yet, what does logical start time really mean? Because, I'm looking at how it's done for the time scheduler, and it's not either the time of execution, it's the time at which the time matched the cron of the schedule. They could be slightly different, right?

anew replied:
ok

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        } catch (TaskExecutionException e) {
+          LOG.error("Execution exception while running streamSizeSchedule {}", currentSchedule.getName(), e);
+          if (e.isRefireImmediately()) {
+            LOG.info("Retrying execution for streamSizeSchedule {}", currentSchedule.getName());
```
So if isRefireImmediately, then you would loop infinitely? In what situations is it true? I think that you should always bail out after a few attempts.

jguery replied:
It's actually always false, as set by ``ScheduleTaskRunner.run``.

anew replied:
ok, as agreed let's remove the loop

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+          throw new IOException("Should collect exactly one time value");
+        }
+        TimeValue timeValue = timeValues.get(0);
+        // The metric store gives us 0 as the timestamp, hence we cannot use it here
```
can that be fixed? Please open a Jira

jguery replied:
I'm not sure if the metrics team would consider that broken, or if that's done on purpose. I had a chat with @gsps1 about getting the timestamp of the last seen metric in that ``TimeValue``, but that's not how it works apparently.

anew replied:
I still think this warrants a Jira. It is a functionality that we want, and the metrics system should be requirement driven. Metrics team can always comment on the Jira and explain why that request makes no sense.

jguery replied:
Created https://issues.cask.co/browse/CDAP-1676.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        activeTasksNow = activeTasks.incrementAndGet();
+      }
+      if (activeTasksNow == 1) {
+        // There were no active tasks until then, that means polling the stream was disabled.
```
where does polling the stream get disabled? suspendScheduleTask does not do that.

jguery replied:
There is no implicit disabling of the polling. This is what happens:

- Last task is suspended, there is 0 active tasks
- The runnable created in ``createPollingRunnable`` is called, but it does nothing since the condition ``if (activeTasks.get() > 0)`` is not met.
- The ``received`` method of this StreamSubscriber is hence not called, which means ``cancelPollingAndScheduleNext`` will not be called again, and we don't reschedule the polling

So exactly one polling should happen after suspending all tasks.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+          }
+        }
+      }
+      task.received(lastNotification);
```
Another remark. If you had two tasks that were both suspended, then you resume both of them. The first one will see activeTasksNow==1, poll for stream size and trigger the task. But the second one sees activeTasksNow==2, and it will not poll. Is there any other way to determine whether you have to poll when the schedule is resumed? Or perhaps you should always do it? Thoughts?

jguery replied:
I think that's what we want here, only one task re-activates the polling. This task will recompute a notification using the polled size, and the second task will be sent that notification too (because ``task.received`` is outside of the ``if (activeTasksNow == 1)`` block).

There could one problem though, if we have the following:
- Task1 resumes, sees activeTasksNow == 1
- Task2 resumes, sees activeTasksNow == 2
- Task2 doesn't go into condition, and executes ``task.received`` with the oldest ``lastNotification``
- Task1 enters the condition, and recomputes ``lastNotification``
- Task1 executes ``task.received`` with the new ``lastNotification``

So depending on when things execute, Task2 may not benefit from the new computed ``lastNotification`` that Task1 is supposed to provide. Again, I thought I could avoid synchronizing the method completely, but it doesn't look like it...

jguery replied:
That also means, in the situation I described, that if ``lastNotification`` is null prior to those resume actions, then Task2 receives a null ``lastNotification``. So I'll just make the method synchronized, it will avoid those issues.

anew replied:
Just combine the two sync blocks.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+     */
+    private void cancelPollingAndScheduleNext() {
+      // This method might be called from the call to #received defined in the below Runnable - in which case
+      // this scheduledPolling would in fact be active. Hence we don't want to interrupt the active task
```
this comment is a little confusing. Because you do interrupt it??

jguery replied:
I will modify the comment, I was referring to the boolean passed to the ``cancel`` method. ``false`` will not interrupt the current ``scheduledPolling``, if it is executing right now.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+    /**
+     * Cancel the currently scheduled stream size polling task, and reschedule one for later.
+     */
+    private void cancelPollingAndScheduleNext() {
```
I think this method must be synchronized. Because it can be invoked concurrently by multiple threads. Both would cancel the existing one, and both would create a new polling runnable and schedule it. Then you will poll twice.

jguery replied:
That's correct, will change.

anew commented:
Hey @jguery I have done another pass. Still some issues... please see my new comments.

jguery commented:
@anew, I addressed your comments and modified the PR as suggested: we now poll the stream every time before triggering a job, so that the logic doesn't have to rely on notifications. Please have another look. Thanks

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    String scheduleId = AbstractSchedulerService.scheduleIdFor(program, programType, scheduleName);
+    StreamSubscriber subscriber = scheduleSubscribers.get(scheduleId);
+    if (subscriber == null) {
+      throw new ScheduleNotFoundException(scheduleName);
```
what happens if you schedule a program, but createScheduleTask() fails (see line 178)? schedule() will return without error, so the schedule should exist. But here you throw a ScheduleNotFound, which seems wrong. Either schedule should error out, or this should throw something else?

jguery replied:
I will make schedule() fail if createScheduleTask() fails.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      } catch (NotFoundException e) {
+        // It can happen that the schedule was deleted while being updated. In which case, the update action
+        // came first and we still want to create it
+        LOG.warn("Schedule {} deleted while being updated", schedule.getName(), e);
```
actually, if the concurrent delete() was successful, then the schedule should be gone. The race condition is whether it gets updated before it gets deleted or not. In both cases, the end result should be that the schedule is deleted. So I think it is ok to propagate the not found exception here. 

More interesting case is if the delete() happened while update is in the middle, that is between this and the next line. In that case, the schedule is temporarily deleted by the update, and delete() will receive a NotFoundException. Caller thinks schedule does not exist, but the truth is that right after it will be recreated. So to be absolutely correct here, we would have to synchronize to prevent delete() to happen once we have entered the if statement. 

I think we can live with that minor race condition, though.

jguery replied:
Ok, so if I propagate the not found exception here, the comment is actually wrong, right?

anew replied:
the comment is wrong. Delete deletes the schedule, regardless of a race with update.

jguery replied:
Ok.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
 
-  private String getScheduleId(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-    return String.format("%s:%s", getProgramScheduleId(program, programType), scheduleName);
+  private synchronized Store getStore() {
```
doesn't it make more sense to create this once in the constructor? Since this store will be used all the time, I don't see a lot of value in lazy instantiation.

jguery replied:
Alright, I'll create the store in the start() then.

anew replied:
yes. Don't we expect that we will trigger execution at least once? If so, then why delay this? It is not expensive.

jguery replied:
Ok, then I'll set the store in the start().

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -74,12 +130,66 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Sch
+    }
+
+    // Add the scheduleTask to the StreamSubscriber
+    if (streamSubscriber.createScheduleTask(program, programType, streamSizeSchedule,
```
not sure when this would return false. One case I know is if the schedule already exists. Then you should throw an exception, not silently ignore the attempted overwrite.

jguery replied:
Right now I make that method return false if the schedule already exists, yes. But you're right I'll change that, and throw an exception instead. The method will return void then.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+                                    newTask);
+        if (previous != null) {
+          // We cannot replace an existing schedule - that functionality is not wanted - yet
+          return false;
```
this should throw an exception. Or the caller should. If I understand correctly, we support updating a schedule, but through a different means.

jguery replied:
Yes, I will throw an exception.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+            cancelPollingAndScheduleNext();
+            try {
+              StreamSize streamSize = queryStreamEventsSize();
+              cancelPollingAndScheduleNext();
```
you just called this 2 lines above? Is that on purpose?

jguery replied:
Oops no this is not on purpose.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        // This is the first time that we schedule this task - ie it was not in the schedule store
+        // before. Hence we set the base info properly by polling the stream
+        try {
+          StreamSize streamSize = queryStreamEventsSize();
```
these two calls seem to come together almost all the time. I feel that the code could be simplified significantly if you restructure it. Let's talk in person

jguery commented:
@anew I simplified the code, as discussed in person. Please have another look. Thank you

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        StreamSize streamSize = pollOnce();
+        sendPollingInfoToActiveTasks(new StreamSizeNotification(streamSize.getTimestamp(), streamSize.getSize()));
+      } catch (Throwable t) {
+        LOG.error("Could not poll stream {}", streamId.getName(), t);
```
The message is not correct as you don't know whether the throwable is from the polling (pollOnce()) or from sending it to the active tasks.

jguery replied:
True, will create multiple catches then.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        send = true;
+        lastNotification = notification;
+      }
+      if (send) {
```
you can simplify this. You don't need the boolean send.

jguery replied:
Right, will remove.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        if (activeTasks.get() > 0) {
+
+          boolean poll = false;
+          if (delta == null) {
```
probably there are two cases: 
- you don't know the delta because you never received a notification before (delta == null)
- the last time you computed the delta is longer ago than the polling interval.

Or can the second case never happen?

jguery replied:
I don't think that we have to deal with the second case here. The logic that triggers a polling in that case will run in ``runOneIteration()``. That logic won't update the delta though.

anew replied:
ok.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+                  LOG.error("Could not poll stream {}", streamId.getName(), t);
+                }
+              }
+            }, Constants.MetricsCollector.DEFAULT_FREQUENCY_SECONDS, TimeUnit.SECONDS);
```
This is the frequency with which the metrics collector emits metrics to the metrics system, right? The actual latency can be a little more, can't it? Have you talked to @abaranau about what is a good delay here?

jguery replied:
I haven't talked with him, but will surely do.

jguery replied:
I talked with him, and he suggested to poll multiple times, still based on this DEFAULT_FREQUENCY_SECONDS.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+                  delta = notification.getSize() - streamSize.getSize();
+                  sendPollingInfoToActiveTasks(new StreamSizeNotification(streamSize.getTimestamp(),
+                                                                          streamSize.getSize()));
+                } catch (Throwable t) {
```
the throwable can be from any of the calls inside the try. The polling might actually have been successful. In that case the message would be misleading.

jguery replied:
Will create multiple catch blocks.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+              @Override
+              public void run() {
+                try {
+                  StreamSize streamSize = pollOnce();
```
can it happen that the regular scheduled polling happens at the same time? In that case, is everything in here safe to do concurrently?

jguery replied:
Yes it can happen, and I think it is safe too. pollOnce() is synchronized, and so the task.reset method won't be called for wrong reasons. It might lead to task.receivedPollingInformation() called concurrently, and the synchronized block in it will make sure that the program is not run twice.

anew replied:
I think you are right

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        }
+
+        if (active) {
+          activeTasks.incrementAndGet();
```
Is this too early to increment it? The task only becomes active after you call startSchedule, which happens after pollOnce(). Seems that for the time being, the number of active tasks becomes inconsistent.

jguery replied:
Yes you're right. It means that I have to include the task.startSchedule() inside the synchronized block too. I will do that.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      if (baseRunSize == -1 && baseRunTs == -1) {
+        newTask.startSchedule(streamSize.getSize(), streamSize.getTimestamp(), active, persist);
+      } else {
+        newTask.startSchedule(baseRunSize, baseRunTs, active, persist);
```
why? This may be a size and time much earlier than the start of this task.

anew replied:
oh, is this the case where baseRunSize/Ts come from the persisted store? So in that case, it would be the tme and size that the task was last triggered?

jguery replied:
I will change the javadoc.

anew replied:
OK for now. Let's revisit this topic when you enable persistence.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+     *
+     * @param pollingInfo {@link StreamSizeNotification} info that came from polling the stream using metrics
+     */
+    public void receivedPollingInformation(@Nonnull StreamSizeNotification pollingInfo) {
```
this should be a StreamSize. Notifications are never used for this, right? It is always from polling.

jguery replied:
Yes, you're right.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      if (task == null) {
+        throw new ScheduleNotFoundException(scheduleName);
+      }
+      if (task.suspend()) {
```
what is the contract for this method? If I try to suspend a schedule that is already suspended, is that an error, or is it silently ignored? Please make sure it is consistent with the time schedule and with the documentation (javadocs and dev guide).

jguery replied:
I checked and for the time schedule, it is not an error to suspend a suspended schedule, it is simply ignored. The javadoc really doesn't mention anything - both ours and quartz's. Our dev guide really doesn't mention anything about schedule lifecycle.
I will modify the javadoc of the ``Scheduler`` interface to specify that this case will be ignored silently.

anew replied:
ok

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      if (task == null) {
+        throw new ScheduleNotFoundException(scheduleName);
+      }
+      if (!task.resume()) {
```
Same as above - is resume a task that isn't suspended an error?

jguery replied:
Same comment as above.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      if (!task.resume()) {
+        return;
+      }
+
```
don't you have to increment the active tasks?

jguery replied:
Yes, I have to that.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      StreamSize streamSize = queryStreamEventsSize();
+      if (lastPollingInfo != null && streamSize.getSize() < lastPollingInfo.getSize()) {
+        for (StreamSizeScheduleTask streamSizeScheduleTask : scheduleTasks.values()) {
+          streamSizeScheduleTask.reset();
```
you cannot reset. That means the task starts over from the beginning of time. You must keep the lastRunTs of the task.

anew replied:
Ok I see further down that reset() sets the time to currentTime. I think it should use the time stamp from the streamSize instead.

jguery replied:
Ok, will use that timestamp.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+     * Send a {@link StreamSizeNotification} built using information from the stream metrics to all the active
+     * {@link StreamSizeSchedule} referenced by this object.
+     */
+    private void sendPollingInfoToActiveTasks(final StreamSizeNotification pollingInfo) {
```
Should be StreamSize

jguery replied:
Will change.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        basePollTs = pollingInfo.getTimestamp();
+
+        argsBuilder.put(ProgramOptionConstants.SCHEDULE_NAME, streamSizeSchedule.getName());
+        argsBuilder.put(ProgramOptionConstants.LOGICAL_START_TIME, Long.toString(pollingInfo.getTimestamp()));
```
these arguments appear redundant. LOGICAL_START_TIME is always the same as RUN_BASE_COUNT_TIME, and same for the tow sizes.

jguery replied:
Well, that's not true, LOGICAL_START_TIME is the time of this run, RUN_BASE_COUNT_TIME is the time that we took as reference to start counting the size of polls, PAST_RUN_LOGICAL_START_TIME is the time at which the previous run was triggered. And the last two can be different, if a reset has occurred between them..

anew replied:
Do you realize that a few lines above, you set basePollSize and basePollTs to whatever is in the pollingInfo? So effectively in the arguments, they are always the same.

jguery replied:
Oh... yes, so I will move the setting of those variables after creating the arguments.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    public void startSchedule(long basePollSize, long basePollTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with basePollSize {}, basePollTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), basePollSize, basePollTs, active, persist);
+      this.basePollSize = basePollSize;
```
So, just to understand this. It seems that basePollSize/Ts is the size and time when the task was last triggered. And lastRunTs/Size is when the task actually was executed. Is that correct? 

If so, then I am not sure whether this initialization is correct. If basePollSize/Ts are -1, then this is a new task that is scheduled for the first time. If not, then they represent what was persisted in the store. So if the task has run before and that was persisted, then why are you setting lastRunTs/Size to -1?

jguery replied:
Yes, you are correct. I will add more arguments to this method, so that we can set both basePoll and lastRun. That means the store will have to persist both those info as well.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      return active.get() && newEstimateSize >= basePollSize + toBytes(streamSizeSchedule.getDataTriggerMB());
+    }
+
+    private long toBytes(int mb) {
```
static

jguery replied:
ok.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+     * Reset the base counters - it happens when polling the stream using metrics shows less data that the previous
+     * poll. This can only be due to metrics deletion or metric TTL.
+     */
+    public void reset() {
```
Oh I see. It does not reset completely.

anew commented:
@jguery looking better. I still have some comments.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+          streamSizeScheduleTask.reset();
+        }
+      }
+      lastPollingInfo = streamSize;
```
do you think you should also set delta to null? Because it is certainly incorrect now.

jguery replied:
That's a good point. We can set it to null here, and in the ``received`` method that deals with notifications, it would be reset right after. And in all the other cases where we call pollOnce, there is no notification involved so you're right, delta becomes incorrect.

jguery commented:
@anew thanks, I have addressed your comments.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java:
```diff
@@ -91,51 +201,628 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+        }
+      }
+      lastPollingInfo = streamSize;
+      delta = null;
```
I meant, not always, but only inside the if statement (if the metrics have decreased)

jguery replied:
Oh, I see. I still think that it's valid to leave delta = null right here. Since when we only do polling, with no notification involved, the delta will not be reset and will also be wrong, unless metrics haven't increased since the last notification received.

anew commented:
I have a few more comments. Most concerning is the loop in the polling. Let's talk together with @abaranau if you need clarification.

anew commented:
LGTM
